src/query/time_of_impact_internal/composite_shape_against_shape.rs

```diff
@@ -39,50 +39,108 @@ where
     composite_shape_against_shape(m2, vel2, g2, m1, vel1, g1)
 }
 
-struct CompositeShapeAgainstAnyTOIVisitor<'a, N: 'a + RealField, G1: ?Sized + 'a> {
-    msum_shift: Vector<N>,
-    msum_margin: Vector<N>,
-    ray: Ray<N>,
-
-    m1: &'a Isometry<N>,
-    vel1: &'a Vector<N>,
-    g1: &'a G1,
-    m2: &'a Isometry<N>,
-    vel2: &'a Vector<N>,
-    g2: &'a Shape<N>,
+/// Time of impact and contact normal of a composite shape with any other shape,
+/// under translational movement.
+pub fn composite_shape_against_shape_with_normal<N, G1: ?Sized>(
+    m1: &Isometry<N>,
+    vel1: &Vector<N>,
+    g1: &G1,
+    m2: &Isometry<N>,
+    vel2: &Vector<N>,
+    g2: &Shape<N>,
+) -> Option<(N, Vector<N>)>
+where
+    N: RealField,
+    G1: CompositeShape<N>,
+{
+    let mut visitor = CompositeShapeAgainstAnyTOIVisitorWithNormal::new(m1, vel1, g1, m2, vel2, g2);
+
+    g1.bvh().best_first_search(&mut visitor)
 }
 
-impl<'a, N, G1: ?Sized> CompositeShapeAgainstAnyTOIVisitor<'a, N, G1>
+/// Time of impact and contact normal of any shape with a composite shape, under
+/// translational movement.
+pub fn shape_against_composite_shape_with_normal<N, G2: ?Sized>(
+    m1: &Isometry<N>,
+    vel1: &Vector<N>,
+    g1: &Shape<N>,
+    m2: &Isometry<N>,
+    vel2: &Vector<N>,
+    g2: &G2,
+) -> Option<(N, Vector<N>)>
 where
     N: RealField,
-    G1: CompositeShape<N>,
+    G2: CompositeShape<N>,
 {
-    pub fn new(
-        m1: &'a Isometry<N>,
-        vel1: &'a Vector<N>,
-        g1: &'a G1,
-        m2: &'a Isometry<N>,
-        vel2: &'a Vector<N>,
-        g2: &'a Shape<N>,
-    ) -> CompositeShapeAgainstAnyTOIVisitor<'a, N, G1>
-    {
-        let ls_m2 = m1.inverse() * m2.clone();
-        let ls_aabb2 = g2.aabb(&ls_m2);
-
-        CompositeShapeAgainstAnyTOIVisitor {
-            msum_shift: -ls_aabb2.center().coords,
-            msum_margin: ls_aabb2.half_extents(),
-            ray: Ray::new(
-                Point::origin(),
-                m1.inverse_transform_vector(&(*vel2 - *vel1)),
-            ),
-            m1: m1,
-            vel1: vel1,
-            g1: g1,
-            m2: m2,
-            vel2: vel2,
-            g2: g2,
+    composite_shape_against_shape_with_normal(m2, vel2, g2, m1, vel1, g1)
```

Comment:
The normal we return should always point toward the outside of the first shape (see my other comment). So it should be reversed here:
```suggestion
    composite_shape_against_shape_with_normal(m2, vel2, g2, m1, vel1, g1).map(|x| (x.0, -x.1))
```